Assignments/Midterm.py

```diff
@@ -0,0 +1,687 @@
+    fig.update_layout(
+        title=f"Continuous Response: {resp} by Continuous Predictor: {pred}",
+        xaxis_title="Predictor",
+        yaxis_title="Response",
+    )
+    file = f"plots/cont_{resp}_cont_{pred}_scatter_plot.html"
+    # fig.show()
+    fig.write_html(
+        file=f"./plots/cont_{resp}_cont_{pred}_scatter_plot.html",
+        include_plotlyjs="cdn",
+    )
+    return file
+
+
+def mean_diff_tbl(pred, resp, data):
+    df = pd.DataFrame(
+        columns=[
+            "(i)",
+            "Bin_Pop",
+            "Bin_Mean",
+            "Pop_Mean",
+            "Pop_Proportion",
+            "Mean_Sq_Diff",
+            "Mean_Sq_Diff_Weighted",
+        ]
+    )
+    pop_mean = data[resp].mean()
+    if not is_continuous(data, pred):
+        for i, val in enumerate(data[pred].unique()):
+            bin_mean = data[resp][data[pred] == val].mean()
+            bin_pop = len(data[resp][data[pred] == val])
+            pop_prop = bin_pop / len(data[resp])
+            mean_sq_diff = (pop_mean - bin_mean) ** 2
+            mean_sq_diff_w = mean_sq_diff * pop_prop
+            df.loc[i] = [
+                val,
+                bin_pop,
+                bin_mean,
+                pop_mean,
+                pop_prop,
+                mean_sq_diff,
+                mean_sq_diff_w,
+            ]
+    else:
+        data_sorted = data.sort_values(pred)
+        data_sorted["bins"] = pd.cut(data_sorted[pred], 10)
+        for i, cur_bin in enumerate(data_sorted["bins"].unique()):
+            bin_mean = data_sorted[resp][data_sorted["bins"] == cur_bin].mean()
+            bin_pop = len(data_sorted[pred][data_sorted["bins"] == cur_bin])
+            pop_prop = bin_pop / len(data_sorted[pred])
+            mean_sq_diff = (pop_mean - bin_mean) ** 2
+            mean_sq_diff_w = mean_sq_diff * pop_prop
+            df.loc[i] = [
+                i,
+                bin_pop,
+                bin_mean,
+                pop_mean,
+                pop_prop,
+                mean_sq_diff,
+                mean_sq_diff_w,
+            ]
+
+    return df
+
+
+def brute_force(pred1, pred2, resp, data):
+    # df = pd.DataFrame(
+    #     columns=[
+    #         "x_1",
+    #         "x_2",
+    #         "Bin_Pop",
+    #         "Bin_Mean",
+    #         "Pop_Mean",
+    #         "Pop_Proportion",
+    #         "Mean_Sq_Diff",
+    #         "Mean_Sq_Diff_Weighted",
+    #     ]
+    # )
+    copy_data = data.copy()
+    if is_continuous(data, pred1):
+        copy_data["bins1"] = pd.cut(copy_data[pred1], 10, labels=False)
+        if is_continuous(data, pred2):
+            copy_data["bins2"] = pd.cut(copy_data[pred2], 10, labels=False)
+            calc_matrix = np.zeros((11, 11))
+            for x_1, val1 in enumerate(copy_data["bins1"].unique()):
+                for x_2, val2 in enumerate(copy_data["bins2"].unique()):
+                    bin_mean = copy_data[resp][
```

Comment:
for continuous data I think it would be easier here to try using pandas groupby function.